#07/04/2022 dictionary comprehension, namedtuples and its advantages in terms of memory

#Dictionary Comprehension

In [ ]:
#create a dictionary from a list
l1 = ["a","b","c","d","e","f"]
{i:i for i in l1}

{'a': 'a', 'b': 'b', 'c': 'c', 'd': 'd', 'e': 'e', 'f': 'f'}

In [ ]:
l1 = ["a","b","c","d","e","f"]
#It creates a dictionary enumerating every item of the list
{i:j for i,j in enumerate(l1)}

{0: 'a', 1: 'b', 2: 'c', 3: 'd', 4: 'e', 5: 'f'}

In [ ]:
sentence = 'the rocket came back from mars'
vowels = [i for i in sentence if i in "aeiou"]
vowels

['e', 'o', 'e', 'a', 'e', 'a', 'o', 'a']

In [ ]:
sentence = 'the rocket came back from mars'
vowels = {i for i in sentence if i in "aeiou"}
vowels

{'a', 'e', 'o'}

In [ ]:
sentence = 'the rocket came back from mars'
vowels = [i for i in sentence if i in "aeiou"]
vowels

['e', 'o', 'e', 'a', 'e', 'a', 'o', 'a']

In [ ]:
quote = "life, uh, finds a way"
unique_vowels = {i for i in quote if i in 'aeiou'}
unique_vowels

{'a', 'e', 'i', 'u'}

#special sorting

In [ ]:
#when you are sorting strings, case matters
#sorted() can be used on a list of strings to sort the values in ascending order, which appears to be alphabetically by

names = ['Harry','Suzy', 'Al', 'Mark']
print(sorted(names))
names_with_case = ['harry','suzy','al','mark']
print(sorted(names_with_case))
names_with_case1 = ['harry','Suzy','al','Mark']
print(sorted(names_with_case1)) #due to unicode answer will be per the thrid print

['Al', 'Harry', 'Mark', 'Suzy']
['al', 'harry', 'mark', 'suzy']
['Mark', 'Suzy', 'al', 'harry']


In [ ]:
#https://unicode.org/emoji/charts/full-emoji-list.html
#the above is unicode for "smileys"
#UNICODE is useful for smileys such type of special characters, where ASCII is not properly suitable. 

In [ ]:
names = ['Harry','Suzy', 'Al', 'Mark']
sorted(names)

['Al', 'Harry', 'Mark', 'Suzy']

In [ ]:
#sorting with key arguments
words = ['banana','pie','washington','book']
sorted(words,key=len)

['pie', 'book', 'banana', 'washington']

In [ ]:
#sorting by reversing. all words are reversed and basing on that sorting process 
words = ['banana','pie','washington','book']
sorted(words,key=lambda x: x[::-1])

['banana', 'pie', 'book', 'washington']

In [ ]:
words = ['banana','pie','washington','book']
sorted(words,key=lambda x: x[::-1], reverse=True)

['washington', 'book', 'pie', 'banana']

In [ ]:
#python sorted() function examples 
l1 = [(2,15),(3,5),(65,5),(8,5)]
#calling function
lisorted = sorted(l1,key=lambda x:sum(x))  #sorting list by getting sum of tuples
print(lisorted)

[(3, 5), (8, 5), (2, 15), (65, 5)]


In [ ]:
from operator import itemgetter
students = {
    0: {"name":"jane","age":40},
    1: {"name":"pool","age":11},
    2: {"name":"dave","age":28}
}

#sort the sub-dicts by age and print
for student in sorted(students.values(),key=itemgetter("age")):
  print(student)

{'name': 'pool', 'age': 11}
{'name': 'dave', 'age': 28}
{'name': 'jane', 'age': 40}


In [ ]:
students.values()

dict_values([{'name': 'jane', 'age': 40}, {'name': 'pool', 'age': 11}, {'name': 'dave', 'age': 28}])

In [ ]:
student.values()

dict_values(['jane', 40])

#Python Collections

In [ ]:
#collection of all miscellanious functions into packages. This is useful for converting dictionaries also immutable
from collections import namedtuple


In [ ]:
a = namedtuple('courses','name, tech')
s = a('data science','python')
print(s)

courses(name='data science', tech='python')


In [ ]:
#how to create a namedtouple using a list?
s._make(['data science','python'])
#the output will be same as before

courses(name='data science', tech='python')

In [ ]:
s.name

'data science'

In [ ]:
s[0]

'data science'

In [ ]:
#this is to make dictionaries ordered
from collections import namedtuple
 
s._asdict()

OrderedDict([('name', 'data science'), ('tech', 'python')])

In [ ]:
a = s._asdict()
print(a)
a['name']

OrderedDict([('name', 'data science'), ('tech', 'python')])


'data science'

In [ ]:
import collections
d = collections.OrderedDict()
d['foo'] = 'python'
d['bar'] = 'span'
d

OrderedDict([('foo', 'python'), ('bar', 'span')])

In [ ]:
d.items()

odict_items([('foo', 'python'), ('bar', 'span')])

In [ ]:
items = list(d.items())
items[0]

('foo', 'python')

In [36]:
#the field vlaues are passed as a string seperated by
from collections import namedtuple
movie = namedtuple('genre','genre rating lead_actor')

#create instance of movie 
ironman = movie(genre='action',rating=8.5,lead_actor='robert downey junior')
titanic = movie(genre='romance',rating=8,lead_actor='leonardo decaprio')
seven = movie(genre='crime',rating=9,lead_actor='Brad pitt')

#access the fields 
print(titanic.genre)
print(seven.lead_actor)
print(ironman.rating)
print(ironman)
movie

romance
Brad pitt
8.5
genre(genre='action', rating=8.5, lead_actor='robert downey junior')


__main__.genre

In [42]:
#why use namedtouple over dictionary
#A major advantage of namedtouple is they take up less space/memory than an equivalent dictionary.
#So, in the case of large data,namedtuples are efficient

#create a dict and namedtuple with same data and compare the size 
import random
import sys
#create dict
dicts = {'numbers_1': random.randint(0, 10000),'numbers_2':random.randint(5000, 10000)}
print('size or space occupied by dictionary',sys.getsizeof(dicts))

#creating same dictionary to nmetuple
data = namedtuple('data',['numbers_1','numbers_2'])
my_namedtuple = data(**dicts)
print('Size or space occupied by namedtuple',sys.getsizeof(my_namedtuple))

size or space occupied by dictionary 248
Size or space occupied by namedtuple 72


#In the above example it is evident that namedtuple is occupying just 72 memory size when compared to regular 248 memory size. 

#CHAIN MAP